Subapp routing using add_domain() does not route as expected.

### Describe the bug

I am opening this issue to link with the xfail test I have created showing this issue.

Adding a subapp with its own routes, with a different domain using app.add_domain(), leads instead to the main app's route being served despite the request.headers[‘host’] value matching the domain provided.

### To Reproduce

```
async def test_subapp_domain_routing(aiohttp_client: AiohttpClient) -> None:
    app = web.Application()
    subApp = web.Application()

    class MyViewWrong(web.View):
        async def get(self) -> web.Response:
            return web.Response(status=404)
    
    class MyViewCorrect(web.View):
        async def get(self) -> web.Response:
            return web.Response()

    app.router.add_routes([web.view("/", MyViewWrong)])
    subApp.router.add_routes([web.view("/", MyViewCorrect)])
    
    app.add_domain('different.example.com', subApp)

    client = await aiohttp_client(app)
    headers = {'Host': 'different.example.com'}
    async with client.get("/", headers=headers) as r:
        assert r.status == 200
```

### Expected behavior

The expected behaviour is that processing will be passed to the registered subapp for it's routem instead of being handled by the main apps route.

### Logs/tracebacks

```python-traceback
/home/me/aiohttp-1/tests/test_web_urldispatcher.py::test_subapp_domain_routing[pyloop] failed: aiohttp_client = <function aiohttp_client.<locals>.go at 0x775d3d3b56c0>

    async def test_subapp_domain_routing(aiohttp_client: AiohttpClient) -> None:
        app = web.Application()
        subApp = web.Application()
    
        class MyViewWrong(web.View):
            async def get(self) -> web.Response:
                return web.Response(status=404)
    
        class MyViewCorrect(web.View):
            async def get(self) -> web.Response:
                return web.Response()
    
        app.router.add_routes([web.view("/", MyViewWrong)])
        subApp.router.add_routes([web.view("/", MyViewCorrect)])
    
        app.add_domain('different.example.com', subApp)
    
        client = await aiohttp_client(app)
        headers = {'Host': 'different.example.com'}
        async with client.get("/", headers=headers) as r:
>           assert r.status == 200
E           AssertionError: assert 404 == 200
E            +  where 404 = <ClientResponse(http://127.0.0.1:36377/) [404 Not Found]>\n<CIMultiDictProxy('Content-Length': '0', 'Date': 'Thu, 16 Oct 2025 05:35:55 GMT', 'Server': 'Python/3.13 aiohttp/4.0.0a2.dev0')>\n.status

MyViewCorrect = <class 'test_web_urldispatcher.test_subapp_domain_routing.<locals>.MyViewCorrect'>
MyViewWrong = <class 'test_web_urldispatcher.test_subapp_domain_routing.<locals>.MyViewWrong'>
aiohttp_client = <function aiohttp_client.<locals>.go at 0x775d3d3b56c0>
app        = <Application 0x775d3d3e5220>
client     = <aiohttp.test_utils.TestClient object at 0x775d3d7f2f90>
headers    = {'Host': 'different.example.com'}
r          = <ClientResponse(http://127.0.0.1:36377/) [404 Not Found]>
<CIMultiDictProxy('Content-Length': '0', 'Date': 'Thu, 16 Oct 2025 05:35:55 GMT', 'Server': 'Python/3.13 aiohttp/4.0.0a2.dev0')>

subApp     = <Application 0x775d3d3e52d0>

tests/test_web_urldispatcher.py:973: AssertionError
```

### Python Version

```console
$ python --version
Python 3.13.7
```

### aiohttp Version

```console
$ python -m pip show aiohttp
Name: aiohttp
Version: 3.13.0
Summary: Async http client/server framework (asyncio)
Home-page: https://github.com/aio-libs/aiohttp
Author: 
Author-email: 
License: Apache-2.0 AND MIT
Location: /home/me/aiohttp-1/.venv/lib/python3.13/site-packages
Requires: aiohappyeyeballs, aiosignal, attrs, frozenlist, multidict, propcache, yarl
Required-by:
```

### multidict Version

```console
$ python -m pip show multidict
Name: multidict
Version: 6.7.0
Summary: multidict implementation
Home-page: https://github.com/aio-libs/multidict
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache License 2.0
Location: /home/me/aiohttp-1/.venv/lib/python3.13/site-packages
Requires: 
Required-by: aiohttp, yarl
```

### propcache Version

```console
$ python -m pip show propcache
Name: propcache
Version: 0.4.1
Summary: Accelerated property cache
Home-page: https://github.com/aio-libs/propcache
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache-2.0
Location: /home/me/aiohttp-1/.venv/lib/python3.13/site-packages
Requires: 
Required-by: aiohttp, yarl
```

### yarl Version

```console
$ python -m pip show yarl
Name: yarl
Version: 1.22.0
Summary: Yet another URL library
Home-page: https://github.com/aio-libs/yarl
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache-2.0
Location: /home/me/aiohttp-1/.venv/lib/python3.13/site-packages
Requires: idna, multidict, propcache
Required-by: aiohttp
```

### OS

Distributor ID: Ubuntu
Description:    Ubuntu 24.04.3 LTS
Release:        24.04
Codename:       noble

### Related component

Server

### Additional context

I was writing a server application with two subapps that each have a single route on "/", and one main app that also has a route on "/". I am using a reverse proxy, but the host and x-forwarded-host all match the domain used in add_domain() for the subroutes when logged by the main app's route. All of this happens in the tests as well, disregarding my specific context. I am going to make a PR with this xfail test and reference this issue.

### Code of Conduct

- [x] I agree to follow the aio-libs Code of Conduct

--- a/CHANGES/11673.bugfix.rst
+++ b/CHANGES/11673.bugfix.rst
@@ -1,0 +1,2 @@
+Fixed routing to a sub-application added via ``.add_domain()`` not working
+if the same path exists on the parent app. -- by :user:`Dreamsorcerer`.

--- a/aiohttp/web_urldispatcher.py
+++ b/aiohttp/web_urldispatcher.py
@@ -977,6 +977,21 @@ class UrlDispatcher(AbstractRouter, Mapping[str, AbstractResource]):
         resource_index = self._resource_index
         allowed_methods: set[str] = set()
 
+        # MatchedSubAppResource is primarily used to match on domain names
+        # (though custom rules could match on other things). This means that
+        # the traversal algorithm below can't be applied, and that we likely
+        # need to check these first so a sub app that defines the same path
+        # as a parent app will get priority if there's a domain match.
+        #
+        # For most cases we do not expect there to be many of these since
+        # currently they are only added by `.add_domain()`.
+        for resource in self._matched_sub_app_resources:
+            match_dict, allowed = await resource.resolve(request)
+            if match_dict is not None:
+                return match_dict
+            else:
+                allowed_methods |= allowed
+
         # Walk the url parts looking for candidates. We walk the url backwards
         # to ensure the most explicit match is found first. If there are multiple
         # candidates for a given url part because there are multiple resources
@@ -993,21 +1008,6 @@ class UrlDispatcher(AbstractRouter, Mapping[str, AbstractResource]):
             if url_part == "/":
                 break
             url_part = url_part.rpartition("/")[0] or "/"
-
-        #
-        # We didn't find any candidates, so we'll try the matched sub-app
-        # resources which we have to walk in a linear fashion because they
-        # have regex/wildcard match rules and we cannot index them.
-        #
-        # For most cases we do not expect there to be many of these since
-        # currently they are only added by `add_domain`
-        #
-        for resource in self._matched_sub_app_resources:
-            match_dict, allowed = await resource.resolve(request)
-            if match_dict is not None:
-                return match_dict
-            else:
-                allowed_methods |= allowed
 
         if allowed_methods:
             return MatchInfoError(HTTPMethodNotAllowed(request.method, allowed_methods))

--- a/docs/web_reference.rst
+++ b/docs/web_reference.rst
@@ -1539,6 +1539,14 @@ Application and Router
       matches the pattern *domain* then
       further resolving is passed to *subapp*.
 
+      .. warning::
+
+         Registering many domains using this method may cause performance
+         issues with handler routing. If you have a substantial number of
+         applications for different domains, you may want to consider
+         using a reverse proxy (such as Nginx) to handle routing to
+         different apps, rather that registering them as sub-applications.
+
       :param str domain: domain or mask of domain for the resource.
 
       :param Application subapp: nested application.

--- a/tests/test_web_urldispatcher.py
+++ b/tests/test_web_urldispatcher.py
@@ -951,8 +951,8 @@ async def test_url_with_many_slashes(aiohttp_client: AiohttpClient) -> None:
         assert r.status == 200
 
 
-@pytest.mark.xfail(reason="https://github.com/aio-libs/aiohttp/issues/11665")
 async def test_subapp_domain_routing_same_path(aiohttp_client: AiohttpClient) -> None:
+    """Regression test for #11665."""
     app = web.Application()
     sub_app = web.Application()
 

